File: python/tests/test_progress.py
"""Tests for pi0.progress: the two output forms and the throttle, without a TTY."""

from __future__ import annotations

import io

from pi0.progress import Progress, _fmt, track


def test_batch_form_is_one_grep_friendly_line():
    s = _fmt("skim", 2000, 5000, 320.0, interactive=False)
    assert "\r" not in s
    assert "40%" in s and "2000/5000" in s and "eta" in s


def test_interactive_form_redraws_in_place_with_a_bar():
    s = _fmt("skim", 2500, 5000, 100.0, interactive=True)
    assert s.startswith("\r")
    assert "#" in s and "50%" in s


def test_zero_total_does_not_divide_by_zero():
    assert "0%" in _fmt("x", 10, 0, 1.0, interactive=False)


def test_percentage_is_clamped():
    assert "100%" in _fmt("x", 7000, 5000, 1.0, interactive=False)


def test_batch_mode_emits_on_a_coarse_step_not_every_update():
    buf = io.StringIO()
    t = [1000.0]  # frozen clock: isolate the 5% percentage trigger from the 30s timer
    p = Progress("load", 100, stream=buf, interactive=False, clock=lambda: t[0])
    for i in range(1, 101):
        p.set(i)
    p.finish()
    lines = buf.getvalue().count("\n")
    assert 18 <= lines <= 25, f"expected ~20 buckets, got {lines}"


def test_interactive_throttles_to_about_ten_frames_a_second():
    buf = io.StringIO()
    t = [0.0]
    p = Progress("run", 1000, stream=buf, interactive=True, clock=lambda: t[0])
    for i in range(1, 101):
        t[0] += 0.01
        p.set(i * 10)
    p.finish()
    frames = buf.getvalue().count("\r")
    assert 8 <= frames <= 14, f"expected ~10 frames, got {frames}"


def test_finish_always_draws_the_final_state():
    buf = io.StringIO()
    t = [0.0]
    p = Progress("run", 10, stream=buf, interactive=False, clock=lambda: t[0])
    p.set(10)
    p.finish()
    assert "100%" in buf.getvalue() and "10/10" in buf.getvalue()


def test_track_advances_one_per_item():
    buf = io.StringIO()
    t = [0.0]
    seen = list(track(range(5), "x", stream=buf, interactive=False, clock=lambda: t[0]))
    assert seen == [0, 1, 2, 3, 4]
    assert "5/5" in buf.getvalue()


def test_add_is_thread_safe():
    import threading

    buf = io.StringIO()
    t = [0.0]
    p = Progress("run", 4000, stream=buf, interactive=False, clock=lambda: t[0])

    def worker():
        for _ in range(1000):
            p.add()

    threads = [threading.Thread(target=worker) for _ in range(4)]
    for th in threads:
        th.start()
    for th in threads:
        th.join()
    p.finish()
    assert "4000/4000" in buf.getvalue()

File: tests/test_progress.cpp
// Unit tests for the progress reporter's pure logic: the two output forms, and
// the throttle that keeps a batch log from filling with redraw frames.
//
// The terminal behaviour itself (carriage returns, isatty) is exercised by the
// end-to-end runs; here we test what can be tested without a TTY, using an
// injected clock and a forced interactive/batch mode.

#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_string.hpp>

#include <sstream>
#include <string>

#include "util/Progress.hpp"

using Catch::Matchers::ContainsSubstring;
using pi0::format_progress;
using pi0::Progress;

TEST_CASE("batch form is a single grep-friendly line, no carriage return", "[progress]") {
    const std::string s = format_progress("skim", 2000, 5000, 320.0, /*interactive=*/false);
    CHECK(s.find('\r') == std::string::npos);
    CHECK_THAT(s, ContainsSubstring("40%"));
    CHECK_THAT(s, ContainsSubstring("2000/5000"));
    CHECK_THAT(s, ContainsSubstring("eta"));
}

TEST_CASE("interactive form redraws in place with a bar", "[progress]") {
    const std::string s = format_progress("skim", 2500, 5000, 100.0, /*interactive=*/true);
    CHECK(s.front() == '\r');           // returns to line start
    CHECK_THAT(s, ContainsSubstring("#"));   // has a bar
    CHECK_THAT(s, ContainsSubstring("50%"));
}

TEST_CASE("percentage is clamped and a zero total does not divide by zero", "[progress]") {
    CHECK_THAT(format_progress("x", 10, 0, 1.0, false), ContainsSubstring("0%"));   // total 0
    CHECK_THAT(format_progress("x", 7000, 5000, 1.0, false), ContainsSubstring("100%"));  // over
    CHECK_THAT(format_progress("x", -5, 5000, 1.0, true), ContainsSubstring("0%"));  // under
}

TEST_CASE("batch mode emits on a coarse step, not on every update", "[progress]") {
    // A frozen clock, so nothing is emitted on the 30s timer -- only the 5%
    // percentage steps fire. 100 updates over a total of 100 should therefore
    // produce ~20 lines (one per 5% bucket), not 100.
    std::ostringstream os;
    double t = 1000.0;  // never advances: isolate the percentage trigger
    Progress p("load", 100, os, /*interactive=*/0, [&] { return t; });
    for (int i = 1; i <= 100; ++i) p.set(i);
    p.finish();
    std::size_t lines = 0;
    for (char ch : os.str()) if (ch == '\n') ++lines;
    INFO("emitted:\n" << os.str());
    CHECK(lines >= 18);   // ~20 buckets
    CHECK(lines <= 25);   // definitely not ~100
}

TEST_CASE("interactive mode throttles to about ten frames a second", "[progress]") {
    // Advance the clock 10 ms per update over 1 s of updates (100 of them). At a
    // 100 ms throttle that is ~10 redraws, plus the forced first and the finish.
    std::ostringstream os;
    double t = 0.0;
    Progress p("run", 1000, os, /*interactive=*/1, [&] { return t; });
    for (int i = 1; i <= 100; ++i) {
        t += 0.01;
        p.set(i * 10);
    }
    p.finish();
    // Count carriage returns as redraw frames.
    std::size_t frames = 0;
    for (char ch : os.str()) if (ch == '\r') ++frames;
    INFO("frames: " << frames);
    CHECK(frames >= 8);
    CHECK(frames <= 14);
}

TEST_CASE("finish always draws the final state even if throttled", "[progress]") {
    std::ostringstream os;
    double t = 0.0;
    Progress p("run", 10, os, /*interactive=*/0, [&] { return t; });
    p.set(10);        // 100%, but the clock has not moved and no bucket may have flushed
    p.finish();
    CHECK_THAT(os.str(), ContainsSubstring("100%"));
    CHECK_THAT(os.str(), ContainsSubstring("10/10"));
}

TEST_CASE("add() accumulates the same as an equivalent set()", "[progress]") {
    std::ostringstream os;
    double t = 0.0;
    Progress p("run", 100, os, /*interactive=*/0, [&] { return t; });
    for (int i = 0; i < 40; ++i) p.add(1);
    p.finish();
    CHECK_THAT(os.str(), ContainsSubstring("40/100"));
}
